# Function for calculatintg Logit (LRT) p-values and the corresponding  
# p-values expected under the null (obtained by permutation of case/control lables)

# Started: Alexey Larionov, 23Jun2019
# Last updated: Alexey Larionov, 23Jun2019

# Arguments:

# "genotypes.mx.mx" is a matrix of additive genotypes.mx (NA,0,1,2) 
# with genes in rows, samples in columns

# "phenotypes.df" is data frame with phenotypes
# It has to contain a "cc" column with case/control lables (0,1)
# all other columns will be included as covariates to the model

# "n_perm" is the number of required permutations

# "show_progress" is only intended for interactive use, requires library svMisc

# Takes ~... min to permute 1000 times a matrix of 2,000 variants x 500 samples

# Consider additional checks with source data and intolerance to warnings ...

logit_variant_qq_perm <- function(genotypes.mx, phenotypes.df, n_perm=1000, show_progress=F){

  # -------------------------------------------------------------- #
  #                           Environment                          #
  # -------------------------------------------------------------- #
  
  # Requires tryCatchAdv
  #source("path/to/f01_tryCatchAdv.R")
  
  # To show progress (not used by default)
  if(show_progress){library(svMisc)}

  # -------------------------------------------------------------- #
  #                     Impute missed genotypes                    #
  # -------------------------------------------------------------- #
  
  # For each variant missed genotypes are substitured by the mean genotype in this variant
  # consistent with "fixed" method of imputation in SKAT
  
  # Prepare matrix for imputed genotypes
  imputed_genotypes.mx <- genotypes.mx
  
  # Impute missed genotypes
  for(var in 1: nrow(genotypes.mx)){
    
    mean_genotype <- mean(genotypes.mx[var,],na.rm=T)
    missed_genotypes <- is.na(genotypes.mx[var,])
    mean_genotype -> imputed_genotypes.mx[var,missed_genotypes]
    
  }
  
  # Clean-up
  rm(var, mean_genotype, missed_genotypes)

  # Show progress if requested (requires library svMisc, not used by default)
  if(show_progress) cat("Completed imputation\n")
  
  # -------------------------------------------------------------- #
  # Function to get sorted vector of p-values for a given additive #
  #          genotypes matrix and a given phenotypes data          #
  # -------------------------------------------------------------- #
  
  get_beta_and_p_values <- function(geno.mx,  pheno.df){
    
    # Fit null model
    null_model <- glm(cc~., data=pheno.df, family=binomial())

    # Get number of variants
    num_var <- nrow(geno.mx)
    
    # Make empty vectors for odd-ratios and p-values for each variant
    glm_p <- rep(NA, num_var)
    glm_beta <- rep(NA, num_var)
    glm_status <- rep(NA, num_var)
    glm_message <- rep(NA, num_var)
    
    # For each variant (row) in the genotypes.mx matrix
    for(var in 1:num_var){
      
      # Initialise all local variables to NA
      glm_p_value <- NA
      glm_beta_value <- NA
      glm_status_value <- NA
      glm_message_value <- NA
      genotypes <- NA
      full_data <- NA
      anova_lrt <- NA
      
      # Get vector of genotypes for the variant
      genotypes <- geno.mx[var,]
      
      # Compile data for full model
      full_data <- cbind(pheno.df, genotypes)
      
      # Fit logit regression 
      full_model <- tryCatchAdv( glm(cc~., data=full_data, family=binomial()) )
      
      # Extract glm completion status (success/error/warning) from the list generated by tryCatchAdv()
      glm_status_value <- full_model$status
      
      # Catch message if error/warning  
      if(glm_status_value != "succeeded") glm_message <- full_model$message$message
      
      # If glm completed with success or warning
      if(glm_status_value != "error"){
        
        # Extract glm() results from the list generated by trtCatchAdv()  
        full_model <- full_model$value

        # Extract beta
        glm_beta_value <- coef(full_model)["genotypes"]
        
        # Estimate p-value by comparison of full model with null model by LRT test
        anova_lrt <- anova(null_model, full_model, test="LRT")
        glm_p_value <- anova_lrt[2,"Pr(>Chi)"]
      }  
      
      # Add values to the result vectors
      glm_p_value -> glm_p[var]
      glm_beta_value -> glm_beta[var]
      glm_status_value -> glm_status[var]
      glm_message_value -> glm_message[var]
      
    } # Next variant
    
    # Add names to result vectors
    names(glm_p) <- rownames(genotypes.mx)
    names(glm_beta) <- rownames(genotypes.mx)
    names(glm_status) <- rownames(genotypes.mx)
    names(glm_message) <- rownames(genotypes.mx)
    
    # Sort all result vectors by p-value order
    glm_p <- sort(glm_p)
    
    glm_beta <- glm_beta[names(glm_p)]
    glm_status <- glm_status[names(glm_p)]
    glm_message <- glm_message[names(glm_p)]

    # Return result
    return(list(glm_p=glm_p,
                glm_beta=glm_beta,
                glm_status=glm_status,
                glm_message=glm_message))
  }
  
  # -------------------------------------------------------------- #
  #                      Sorted observed p-values                  #
  # -------------------------------------------------------------- #
  
  observed <- get_beta_and_p_values(imputed_genotypes.mx, phenotypes.df)

  # Show progress if requested (requires library svMisc, not used by default)
  if(show_progress) cat("Calculated p-values for observed data\nStarting permutations ...\n")
  
  # -------------------------------------------------------------- #
  #  Sorted p-values expected under the null (by permuting lables) #
  # -------------------------------------------------------------- #
  
  # Make empty matrix to accumulate p-values during permutations
  p.mx <- matrix(NA, ncol=nrow(imputed_genotypes.mx), nrow=0)
  
  # Accumulate sorted permuted p-values for the required number of permutations
  for(p in 1:n_perm){
    
    # Permute cc - lables using sample() function, 
    permuted_phenotypes.df <- phenotypes.df
    permuted_phenotypes.df$cc <- sample(permuted_phenotypes.df$cc)
    
    # Calculate the vector of logit p-values, local get_beta_and_p_values() function
    # (the vecror has already been sorted), add the vector to permutations matrix
    permuted <- get_beta_and_p_values(imputed_genotypes.mx, permuted_phenotypes.df)
    p.mx <- rbind(p.mx, permuted$glm_p)
    
    # Show progress if requested (requires library svMisc, not used by default)
    if(show_progress) progress(value = p, max.value = n_perm)
    
  } # Next permutation
  
  # Calculate expected p-values as mean (and 95CI) of the permuted p-values
  p_exp <- apply(p.mx,2,mean,na.rm=T)
  p_exp_95pp <- apply(p.mx,2,quantile,probs=0.95,na.rm=T)
  p_exp_05pp <- apply(p.mx,2,quantile,probs=0.05,na.rm=T)
  names(p_exp) <- NULL
  names(p_exp_95pp) <- NULL
  names(p_exp_05pp) <- NULL
  
  # Compile "expected" as a list with the central expectation of p-values and 95CI
  expected <- list(p_values=p_exp,p_exp_05pp=p_exp_05pp,p_exp_95pp=p_exp_95pp)
  
  # Return observed (p, beta, glm status and message) and expected (p)
  return(list(observed=observed, expected=expected))
  
}
